### PR TITLE
Add semantic conventions for Elasticsearch client instrumentation

### PR DESCRIPTION
**Pull request by estolfo**
This PR originated as a [PR in the opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification/pull/3358) repo.
The outstanding discussion point from that PR was regarding the path forward for the existing Elasticsearch client instrumentations:

- .NET: No concerns about updating their instrumentation to adopt these semantic conventions
- Python: No concerns about updating their instrumentation to adopt these semantic conventions
- Java: request to understand the complexity of updating the instrumentation by opening a PR with the changes

We are in the process of figuring out the path forward for the Java instrumentation and I'll update here when we have a plan.

Two instrumentations in progress that will use these semantic conventions:
- PHP
- [Ruby](https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/357)

### CODE REVIEW

File: semantic_conventions/trace/database.yaml
Comment (lmolkova) [inline semantic_conventions/trace/database.yaml:424]:
```suggestion
        examples: [ '"{\"name\":\"TestUser\",\"password\":\"REDACTED\"}"' ]
```

File: semantic_conventions/trace/database.yaml
Comment (lmolkova) [inline semantic_conventions/trace/database.yaml:414]: HTTP client spans have `url.full`, but not the individual components, so would it make sense to populate it? or populate individual parts specific to elastic like `target`, `doc_id`? or both?

File: semantic_conventions/trace/database.yaml
Comment (lmolkova) [inline semantic_conventions/trace/database.yaml:408]: it's a bit confusing to me to see it as `url.query` even if it's how it's passed in case of elastic.

why not put it into `db.statement`, especially since it's already used to record query when it's passed in the body?

Would it make sense to capture `url.full` to comply with HTTP spec and putting DB semantics into `db.*` attributes?

File: semantic_conventions/trace/database.yaml
Comment (lmolkova) [inline semantic_conventions/trace/database.yaml:422]: can we define it as a query and, optionally, if important, add  a note saying that it can be passed as body or as a query string?

File: semantic_conventions/trace/database.yaml
Comment (estolfo) [reply semantic_conventions/trace/database.yaml:414]: I'm not sure if I understand what you are proposing. Can you clarify if you are proposing not to have `url.path` but instead `url.full`? and whether you are proposing to keep the individual components as span attributes?

I think there is some value to having the individual parts as span attributes if, for example, you want to visualize spans that query particular indexes in Elasticsearch. The name of the span should be the path containing placeholders for these individual values, so I think there should be another way to target spans based on the placeholder values.

File: semantic_conventions/trace/database.yaml
Comment (estolfo) [reply semantic_conventions/trace/database.yaml:422]: Sometimes the body is not necessarily a query, like a document(s) to index, so defining it as an elastic query might not make sense.

**Comment by estolfo** [comment]
Hi @lmolkova, we've discussed what changes to make based on your comments/questions.

We agree it makes sense to have `url.full` instead of the `url.path` and `url.query`. We think it's still important to have the components extracted out, so we'll have `doc_id`, `target`, `url.full` and the span name containing the path pattern with placeholders that will together provide a complete set of information on the api queried and with what data.
We've added a note that the server address and port are recommended.

We are also assuming from the spec that there will also be an http span.

Lastly, we've changed the `db.statement` to capture the request body for search-type queries. The list of Search APIs can be found [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/search.html).

A config option could also be used that allows users to opt-in for capturing bodies of every request.

**Review by joaopgrassi — COMMENTED** [review]
It would be great to get some insights on what is the intended span structure such conventions want to achieve. The initial paragraph about an HTTP span already existing is a bit unclear if an additional span for Elasticsearch should always be there, or attributes should be added to the underlying http span.

File: semantic_conventions/trace/database.yaml
Comment (lmolkova) [inline semantic_conventions/trace/database.yaml:418]: nit, feel free to ignore: no need to specify the requirement level here, `recommended` is a default value

**Review by lmolkova — APPROVED** [review]

**Review by lmolkova — APPROVED** [review]
@estolfo thank you for addressing my comments! Please also add a changelog entry.

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: In other conventions, we have certain recommendation/guidelines for which spans should be created (e.g., [CloudEvents](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/cloudevents.md#spans). So I wonder if we shouldn't have something similar here.

At least to me, from the text above it's not clear if instrumentation SHOULD always create a separate span for Elasticsearch client requests or if they should just add the attributes here to any existing HTTP client spans. Wouldn't the lack of such guidelines also create a inconsistency between language instrumentation like we have now?

@estolfo sorry if this was discussed somewhere already, but I couldn't find it.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: Hi @joaopgrassi thanks for your input. I purposely left it vague and didn't offer guidance on whether to create a separate http span because I think it depends on the implementation, what the user wants, and what backend/visualization the user is using.

What this spec is saying is that at least there will be one span with these attributes for an elasticsearch client request. There may or may not be a http span as well, but its existence doesn't change these semantic conventions.

I think the best approach, ultimately, would be to expose a config option to create separate http spans or just have the elasticsearch spans but I'm not sure that's something that should be mentioned in semantic conventions.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: Hi @joaopgrassi again, sorry for the back-and-forth but after talking internally, I think it actually might make sense to recommend that by default, no http span be created. Ultimately, what an Elasticsearch span represents is a db operation so it doesn't seem necessary to also have a network level span as well.
So what do you think about recommending a default of no http spans but suggesting a config for allowing the creation of them?

Also, we thought it might make sense to also add a `db.operation` attribute. This would correspond to the endpoint id, so it would be the "name" field in this spec: https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: Yes that's what I was trying to get at. With the current text I felt it was not clear if one should "mix" attributes with the HTTP span making the request to Elasticsearch or if a new one should always be created to model this "logical operation". 

@estolfo 

> I think it actually might make sense to recommend that by default, no http span be created

I'm not sure that's a good idea. From what you said and @Oberon00 then I think we simply need to make it more clear that instrumentations SHOULD create a span to model the Elasticsearch call and add the attributes as specified here. Then the question about creating/not creating underlying HTTP spans is up to the user if they add HTTP auto instrumentation or if the underlying HTTP client used by the Elasticsearch client library is also instrumented. What do you think?

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: Yes that's what I was trying to get at. With the current text I felt it was not clear if one should "mix" attributes with the HTTP span making the request to Elasticsearch or if a new one should always be created to model this "logical operation". 

@estolfo 

> I think it actually might make sense to recommend that by default, no http span be created

I'm not sure that's a good idea. From what you said and @Oberon00 then I think we simply need to make it more clear that instrumentations SHOULD create a span to model the Elasticsearch call and add the attributes as specified here. No mention about any other span. 

Then the question about creating/not creating underlying HTTP spans is up to the user if they add HTTP auto instrumentation or if the underlying HTTP client used by the Elasticsearch client library is also instrumented. What do you think?

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: ok, so if I understand the suggestion at this point: the spec should suggest that a span be created for an Elasticsearch request with the attributes described and not mention anything about an http span?

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:6]: Yeah I'd say that would makes the most sense, and if I also got it right, it's the direction @Oberon00  was going at with this previous comment above.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md]: Hi @joaopgrassi, I updated the description to remove a reference to http spans. 
I also added `db.operation` to the span attributes to capture the generic name of the request. It's the "name" field in the json schema [here](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json)

File: semantic_conventions/trace/database.yaml
Comment (joaopgrassi) [inline semantic_conventions/trace/database.yaml:394]: Nit/Non-blocking: I always think it's nice for consuming the conventions when the `Required` attributes come first in the table. When they are "mixed" I find it hard to read/interpret things.

**Review by joaopgrassi — APPROVED** [review]

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md]: I would still prefer some sort of "SHOULD" statement indicating that when talking to Elasticsearch such spans should be created, but I guess being vague is also fine 👍.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md]: I understand but it feels a bit strange to describe that a span should be created in the semantic conventions doc itself-- the existence of the semantic conventions doc implies that a span should be created.

File: semantic_conventions/trace/database.yaml
Comment (joaopgrassi) [reply semantic_conventions/trace/database.yaml:403]: Yeah, the rendered table. Weird I thought the order would follow the yaml. Never mind then :)

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (Mpdreamz) [inline specification/trace/semantic_conventions/instrumentation/elasticsearch.md:23]: `{target}` seems to indicate a singular where it could be plural. Might be good to normalize `type` to `array` (strings).



Maybe we want to generalise this section to: 

`db.elasticsearch.parts.{namefromspec}` 
that way we can pick up url parts from any Elasticsearch call:

https://github.com/search?q=repo%3Aelastic%2Felasticsearch-specification%20path_parts&type=code


That would make this available as `db.elasticsearch.parts.index` as per 
https://github.com/elastic/elasticsearch-specification/blob/main/specification/_global/search/SearchRequest.ts#L60 but we could discuss renaming this part to `target` since for a long time it does not solely apply to indices anymore.

**Review by Mpdreamz — CHANGES_REQUESTED** [review]

**Comment by estolfo** [comment]
@lmolkova and @joaopgrassi I've updated the span name and attributes after a discussion with some more people internally. 
There are two main changes:
- span name is the endpoint id of the request (the `name` field [here](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json), ex: `search`), rather than the http request method and request template (ex `GET /{target}/_doc/{id}`) . The Elasticsearch clients team said people would more likely want to identify ES requests by the endpoint id and that the request template would be less natural for them to see as a request identifier. We also have the http method anyway in the span attributes so we don't need to repeat it in the span name.
- the dynamic parts of the url path are set at span attributes like they are for [http headers](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/http.md#http-request-and-response-headers), as `db.elasticsearch.path_parts.<key>`

What do you think?

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [inline specification/trace/semantic_conventions/instrumentation/elasticsearch.md:27]: I assume this is possible to know from the ElasticSearch client libraries?

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [inline specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: The `key` here is one of the allowed/known "parameters that can be passed to a given ES URL? Are those an unbounded list or do we know all of them? I'm not sure how I feel about this, since it's not listed in YAML but just defined here. 

One clear benefit for ES is if it adds new parameters, then the conventions here don't need to change. 
But in counterpart, if there aren't many of them would it be better to list them all here and have them part of the YAML definitions? 

Instrumentation usually have constants for such attribute key names and if we use this, then they need to construct such attribute keys "all the time", no? 

Also, the values that are recorded here, since they come from the URL, could they contain sensitive info/invalid data that needs to be sanitized? Do you want to always have this recorded or maybe should this be Opt-in?

**Review by joaopgrassi — COMMENTED** [review]
Given you are the subject-matter expert here I think the span name change makes sense (makes more sense to customers). But the `path_parts` change I'm not sure. I left a common with my concerns there

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:27]: Seems so, from https://github.com/open-telemetry/semantic-conventions/pull/23#discussion_r1221535423. all good.

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: Hi @joaopgrassi there is indeed a finite list of the url path parameters. You can see them as the elements in `{ }` in the `path` values of the [spec](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json). They are not runtime dynamic; any path parameters for a given version of ES would be defined in the spec. I don't think it would be a good idea to enumerate them in YAML because 1) there are too many and 2) that would require semantic convention updates with each new release of ES.

The path parameters names themselves would not contain sensitive info, as they only represent keys. The values, however, are either identifiers produced by ES itself or very technical identifiers (e.g. snapshot repositories, template names, etc). The only path value that could potentially contain PII is the document id, as sometimes people put an email address or name there but that's generally bad practice and advised against.

File: semantic_conventions/trace/database.yaml
Comment (Oberon00) [reply semantic_conventions/trace/database.yaml]: (Possibly a bug in the tooling, see https://github.com/open-telemetry/build-tools/issues/133#issuecomment-1428836054)

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: Thanks for the info. The problem about code generation of attribute constants is still a problem though. I don't think there's any language using our build-tools to generate the code files containing the attributes from YAML, but at some point I think they will. With the introduction of this attribute that will not be possible (it will be will render an attribute named `db.elasticsearch.path_parts.<key>`). 

I don't know how to go about this now. Either allow this as is and open a separated issue to discuss it, or go with this PR without it and open a separated issue/PR to discuss.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: I took this approach from [http headers](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/http.md#http-request-and-response-headers)
Is there a plan for that? I would think we should open a separate issue since this and the http headers approach should be aligned.

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: Just to clarify: I think the approach for generating code files from the `part_parts` notation (and http headers notation) shouldn't block the approval and merging of this PR. What do you think?

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: Ah my bad, I somehow thought it was part of the `yaml` file but I see now it's just in the markdown. I guess then all good, instrumentations will have to do their own thing to record these, and since we have precedence all should be good. Thanks for looking into it!

**Review by joaopgrassi — APPROVED** [review]

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: ok, great, thank you for your review!

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:25]: What are the next steps for getting the PR merged?

**Comment by joaopgrassi** [comment]
@open-telemetry/specs-semconv-maintainers  could you please take a look? Thanks!

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (jsuereth) [inline specification/trace/semantic_conventions/instrumentation/elasticsearch.md:35]: This is an interesting conundrum here. There's an OTEP around providing generic "request body" and "response body" attributes on spans.

I think what you've defined here makes a lot of sense in the context of Elastic Search.

Could you provide more guidance on sanitizing sensitive information?   Is it "obvious" what could be sensitive in elastic search from the instrumentation author of the client library, or does this require knowledge of the datamodel in the database?  IF the later, maybe there needs to be recommended configuration options for this attribute.

**Comment by jsuereth** [comment]
Generally looks good to me, but have some concerns around the implications of sensitive data handling.

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:35]: @jsuereth, the request body for an Elasticsearch query could indeed contain sensitive information. Do semantic conventions typically suggest a config option? The Graphql and database `db.statement` fields suggest sanitization but don't mention config options so I'm happy to add something but not sure what type of suggestion.
Graphql: [`The value may be sanitized to exclude sensitive information.`](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/instrumentation/graphql.md?plain=1#L20)
database: [`Should be collected by default only if there is sanitization that excludes sensitive information.`](https://github.com/open-telemetry/semantic-conventions/blob/main/specification/trace/semantic_conventions/database.md?plain=1#L187C10-L187C107)

Comment (jsuereth) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:35]: As discussed, once your notice changes to the database variant, then this PR looks good to go!

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:35]: done! https://github.com/open-telemetry/semantic-conventions/pull/23/commits/1bd79a1cb4aa851df39e421c9ab13a2cfe2e3952

**Review by joaopgrassi — APPROVED** [review]
Depending when this merges, it might be affected by #143

**Comment by estolfo** [comment]
I think @carlosalberto said in the SIG last night that he'd approve it and then we can merge.

**Review by carlosalberto — APPROVED** [review]

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (carlosalberto) [inline specification/trace/semantic_conventions/instrumentation/elasticsearch.md:7]: Minor non-blocking comment: an example could help people writing new instrumentation for Elastic, etc.

**Review by jsuereth — APPROVED** [review]

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (jsuereth) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:7]: This is ready to merge once this comment is acknowledged or addressed :)

File: specification/trace/semantic_conventions/database.md
Comment (joaopgrassi) [inline specification/trace/semantic_conventions/database.md:368]: Given there's a `elasticsearch.md` now, I would move the examples there instead of here. It will already help in the migration that @AlexanderWert in #144

Comment (estolfo) [reply specification/trace/semantic_conventions/database.md]: ok, I've moved it. Thanks!

Comment (joaopgrassi) [reply specification/trace/semantic_conventions/database.md]: Perfect! thank you :)

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (joaopgrassi) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md:7]: Example was added in cbdfd90c10c8527ca9120b8b5aa798cc9b317e33, this is ready to go :shipit:

File: specification/trace/semantic_conventions/instrumentation/elasticsearch.md
Comment (arminru) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md]: @estolfo @Mpdreamz This should be resolved now with the changes in ad963747464a4f947e0751db92b16c5db6e0d692, right?

Comment (estolfo) [reply specification/trace/semantic_conventions/instrumentation/elasticsearch.md]: Yup!

**Review by Mpdreamz — APPROVED** [review]
LGTM

**Pull request merged**
